import os

configfile: "config.yaml"
SAMPLES = config["samples"]

rule all:
    input:
        expand("counts/{sample}_gene_counts.txt", sample=SAMPLES)

rule fastqc_raw:
    input:
        fq1 = "raw_fastq/{sample}_1.fastq",
        fq2 = "raw_fastq/{sample}_2.fastq"
    output:
        "fastqc_raw/{sample}_1_fastqc.html",
        "fastqc_raw/{sample}_2_fastqc.html"
    threads: 2
    singularity: config["singularity_images"]["fastqc"]
    shell:
        "fastqc {input.fq1} {input.fq2} -o fastqc_raw -t {threads}"

rule trimmomatic:
    input:
        fq1 = "raw_fastq/{sample}_1.fastq",
        fq2 = "raw_fastq/{sample}_2.fastq"
    output:
        fq1p = "trimmed_fastq/{sample}_1P.fastq",
        fq1u = "trimmed_fastq/{sample}_1U.fastq",
        fq2p = "trimmed_fastq/{sample}_2P.fastq",
        fq2u = "trimmed_fastq/{sample}_2U.fastq"
    threads: config["trimmomatic"]["threads"]
    singularity: config["singularity_images"]["trimmomatic"]
    shell:
        """
        trimmomatic PE -threads {threads} \
        {input.fq1} {input.fq2} \
        {output.fq1p} {output.fq1u} \
        {output.fq2p} {output.fq2u} \
        ILLUMINACLIP:{config[reference][adapters]}:2:30:10 \
        SLIDINGWINDOW:{config[trimmomatic][slidingwindow]} \
        TRAILING:{config[trimmomatic][trailing]} \
        MINLEN:{config[trimmomatic][minlen]}
        """

rule fastqc_trimmed:
    input:
        fq1 = "trimmed_fastq/{sample}_1P.fastq",
        fq2 = "trimmed_fastq/{sample}_2P.fastq"
    output:
        "fastqc_trimmed/{sample}_1P_fastqc.html",
        "fastqc_trimmed/{sample}_2P_fastqc.html"
    threads: 2
    shell:
        "fastqc {input.fq1} {input.fq2} -o fastqc_trimmed -t {threads}"

rule star_align:
    input:
        fq1 = "trimmed_fastq/{sample}_1P.fastq",
        fq2 = "trimmed_fastq/{sample}_2P.fastq"
    output:
        bam = "star_aligned/{sample}_Aligned.sortedByCoord.out.bam"
    threads: 4
    params:
        prefix = "star_aligned/{sample}_"
    config["singularity_images"]["star"]
    shell:
        """
        STAR --runThreadN {threads} \
             --genomeDir {config[reference][star_index]} \
             --readFilesIn {input.fq1} {input.fq2} \
             --outFileNamePrefix {params.prefix} \
             --outSAMtype BAM SortedByCoordinate
        """

rule featurecounts:
    input:
        bam = "star_aligned/{sample}_Aligned.sortedByCoord.out.bam"
    output:
        "counts/{sample}_gene_counts.txt"
    threads: 4
    config["singularity_images"]["featurecounts"]
    shell:
        """
        featureCounts -T {threads} -p -t exon -g gene_id \
            -a {config[reference][gtf]} \
            -o {output} {input.bam}
        """
